File: MiniDOMBase/DMD/Sistema/Calendar/Providers/BaseCalendarActivitiesProvider.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;


namespace minidom
{


    

    public partial class Sistema
    {

        /// <summary>
        /// Classe base del provider di attività del calendario
        /// Classe base che implementa i metodi dell'interfaccia ICalendarProvider
        /// </summary>
        /// <remarks></remarks>
        public abstract class BaseCalendarActivitiesProvider 
            : ICalendarProvider
        {

            private CModulesClass m_Module;

            /// <summary>
            /// Costruttore
            /// </summary>
            public BaseCalendarActivitiesProvider()
            {
                //DMDObject.IncreaseCounter(this);
                m_Module = null;
            }

            /// <summary>
            /// Restituisce il repository
            /// </summary>
            public CModulesClass Module
            {
                get
                {
                    if (m_Module is null)
                        m_Module = GetModule();
                    return m_Module;
                }
            }

            /// <summary>
            /// Restituisce il repository
            /// </summary>
            /// <returns></returns>
            protected virtual CModulesClass GetModule()
            {
                return null;
            }

            /// <summary>
            /// Istanzia un nuovo oggetto gestito
            /// </summary>
            /// <returns></returns>
            public virtual object InstantiateNewItem()
            {
                return new CCalendarActivity();
            }

            /// <summary>
            /// Restituisce il comando di creazione
            /// </summary>
            /// <returns></returns>
            public virtual string GetCreateCommand()
            {
                string GetCreateCommandRet = default;
                GetCreateCommandRet = "/calendar/activities/?_a=create";
                return GetCreateCommandRet;
            }

            /// <summary>
            /// Restituisce la descrizione del provider
            /// </summary>
            /// <returns></returns>
            public virtual string GetShortDescription()
            {
                return "Attività";
            }

            /// <summary>
            /// Restituisce i tipi gestiti dal provider
            /// </summary>
            /// <returns></returns>
            public virtual Type[] GetSupportedTypes()
            {
                return new[] { typeof(CCalendarActivity) };
            }

            /// <summary>
            /// Restituisce una collezione di oggetti CActivePerson contenente tutte le persone con attività in sospes o nelle date indicate. La collezione è ordinata in funzione della priorità o del ritardo delle operazioni
            /// </summary>
            /// <param name="nomeLista"></param>
            /// <param name="fromDate"></param>
            /// <param name="toDate"></param>
            /// <param name="ufficio"></param>
            /// <param name="operatore"></param>
            /// <returns></returns>
            /// <remarks></remarks>
            public abstract CCollection<CActivePerson> GetActivePersons(
                            string nomeLista, 
                            DateTime? fromDate, 
                            DateTime? toDate, 
                            int ufficio = 0, 
                            int operatore = 0
                            );

            /// <summary>
            /// Restituisce le scadenze nell'intevallo compreso tra fromDate e toDate (incluse)
            /// </summary>
            /// <param name="fromDate"></param>
            /// <param name="toDate"></param>
            /// <returns></returns>
            public abstract CCollection<ICalendarActivity> GetScadenze(DateTime? fromDate, DateTime? toDate);

            /// <summary>
            /// Restituisce le scadenza pendenti (fino ad oggi)
            /// </summary>
            /// <returns></returns>
            public abstract CCollection<ICalendarActivity> GetPendingActivities();

            /// <summary>
            /// Restituisce l'elenco delle cose da fare per l'utente specificato
            /// </summary>
            /// <param name="user"></param>
            /// <returns></returns>
            /// <remarks></remarks>
            public abstract CCollection<ICalendarActivity> GetToDoList(CUser user);

            /// <summary>
            /// Restituisce il nome che identifica univocamente il provider all'interno del sistema
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public abstract string UniqueName { get; }

            /// <summary>
            /// Elimina l'attività
            /// </summary>
            /// <param name="item"></param>
            /// <param name="force"></param>
            public abstract void DeleteActivity(ICalendarActivity item, bool force = false);

            /// <summary>
            /// Salva l'attività
            /// </summary>
            /// <param name="item"></param>
            /// <param name="force"></param>
            public abstract void SaveActivity(ICalendarActivity item, bool force = false);

            ///// <summary>
            ///// Distruttore
            ///// </summary>
            //~BaseCalendarActivitiesProvider()
            //{
            //    //DMDObject.DecreaseCounter(this);
            //}
        }
    }
}

File: MiniDOMBase/DMD/Sistema/Calendar/Providers/BaseCalendarActivitiesProvider.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica


' Classe:		BaseCalendarActivitiesProvider
' Descrizione:	Classe base che implementa i metodi dell'interfaccia ICalendarProvider
' Versione:	    1.0.0.0
' Data:		    20/12/2017

Partial Class Sistema

    ''' <summary>
    ''' Classe base del provider di attività del calendario
    ''' </summary>
    ''' <remarks></remarks>
    Public MustInherit Class BaseCalendarActivitiesProvider
        Implements ICalendarProvider

        Private m_Module As CModule

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.m_Module = Nothing
        End Sub

        Public ReadOnly Property [Module] As CModule
            Get
                If (m_Module Is Nothing) Then m_Module = Me.GetModule
                Return m_Module
            End Get
        End Property

        Protected Overridable Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overridable Function InstantiateNewItem() As Object Implements ICalendarProvider.InstantiateNewItem
            Return New CCalendarActivity
        End Function

        Public Overridable Function GetCreateCommand() As String Implements ICalendarProvider.GetCreateCommand
            GetCreateCommand = "/calendar/activities/?_a=create"
        End Function

        Public Overridable Function GetShortDescription() As String Implements ICalendarProvider.GetShortDescription
            Return "Attività"
        End Function

        Public Overridable Function GetSupportedTypes() As System.Type() Implements ICalendarProvider.GetSupportedTypes
            Return {GetType(CCalendarActivity)}
        End Function

        ''' <summary>
        ''' Restituisce una collezione di oggetti CActivePerson contenente tutte le persone con attività in sospes o nelle date indicate. La collezione è ordinata in funzione della priorità o del ritardo delle operazioni
        ''' </summary>
        ''' <param name="fromDate"></param>
        ''' <param name="toDate"></param>
        ''' <param name="ufficio"></param>
        ''' <param name="operatore"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public MustOverride Function GetActivePersons(ByVal nomeLista As String, ByVal fromDate As Date?, ByVal toDate As Date?, Optional ByVal ufficio As Integer = 0, Optional ByVal operatore As Integer = 0) As CCollection(Of CActivePerson) Implements ICalendarProvider.GetActivePersons

        Public MustOverride Function GetScadenze(ByVal fromDate As Date?, ByVal toDate As Date?) As CCollection(Of ICalendarActivity) Implements ICalendarProvider.GetScadenze

        Public MustOverride Function GetPendingActivities() As CCollection(Of ICalendarActivity) Implements ICalendarProvider.GetPendingActivities

        ''' <summary>
        ''' Restituisce l'elenco delle cose da fare per l'utente specificato
        ''' </summary>
        ''' <param name="user"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public MustOverride Function GetToDoList(ByVal user As CUser) As CCollection(Of ICalendarActivity) Implements ICalendarProvider.GetToDoList

        ''' <summary>
        ''' Restituisce il nome che identifica univocamente il provider all'interno del sistema
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public MustOverride ReadOnly Property UniqueName As String Implements ICalendarProvider.UniqueName

        Public MustOverride Sub DeleteActivity(item As ICalendarActivity, Optional ByVal force As Boolean = False) Implements ICalendarProvider.DeleteActivity

        Public MustOverride Sub SaveActivity(item As ICalendarActivity, Optional ByVal force As Boolean = False) Implements ICalendarProvider.SaveActivity

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Class